libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1681,161 +1681,198 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+							identifier->annotation().referencedDeclaration
+						)
+					)
+						ssl.append(
+							"Did you mean to declare this variable as \"address payable\"?",
+							variableDeclaration->location()
+						);
+				m_errorReporter.typeError(
+					_functionCall.location(), ssl,
+					"Explicit type conversion not allowed from non-payable \"address\" to \"" +
+					resultType->toString() +
+					"\", which has a payable fallback function."
+				);
 			}
+			else
+				m_errorReporter.typeError(
+					_functionCall.location(),
+					"Explicit type conversion not allowed from \"" +
+					argType->toString() +
+					"\" to \"" +
+					resultType->toString() +
+					"\"."
+				);
+		}
+		if (resultType->category() == Type::Category::Address)
+		{
+			bool const payable = argType->isExplicitlyConvertibleTo(AddressType::addressPayable());
+			resultType = make_shared<AddressType>(
+				payable ? StateMutability::Payable : StateMutability::NonPayable
+			);
 		}
-		_functionCall.annotation().type = resultType;
-		_functionCall.annotation().isPure = isPure;
-
-		return false;
 	}
+	return resultType;
+}
 
-	// Actual function call or struct constructor call.
+TypePointer TypeChecker::typeCheckFunctionLikeAndRetrieveReturnType(
+	FunctionCall const& _functionCall,
+	FunctionTypePointer _functionType
+)
+{
+	if (!_functionType)
+	{
+		m_errorReporter.typeError(_functionCall.location(), "Type is not callable");
+		return make_shared<TupleType>();
+	}
 
-	FunctionTypePointer functionType;
+	TypePointers returnTypes;
 
-	/// For error message: Struct members that were removed during conversion to memory.
-	set<string> membersRemovedForStructConstructor;
-	if (_functionCall.annotation().kind == FunctionCallKind::StructConstructorCall)
+	if (_functionType->kind() == FunctionType::Kind::ABIDecode)
 	{
-		TypeType const& t = dynamic_cast<TypeType const&>(*expressionType);
-		auto const& structType = dynamic_cast<StructType const&>(*t.actualType());
-		functionType = structType.constructorType();
-		membersRemovedForStructConstructor = structType.membersMissingInMemory();
-		_functionCall.annotation().isPure = isPure;
+		bool const abiEncoderV2 = m_scope->sourceUnit().annotation().experimentalFeatures.count(
+				ExperimentalFeature::ABIEncoderV2
+			);
+		returnTypes = typeCheckABIDecodeAndRetrieveReturnType(_functionCall, abiEncoderV2);
 	}
-	else if ((functionType = dynamic_pointer_cast<FunctionType const>(expressionType)))
-		_functionCall.annotation().isPure =
-			isPure &&
-			_functionCall.expression().annotation().isPure &&
-			functionType->isPure();
-
-	bool allowDynamicTypes = m_evmVersion.supportsReturndata();
-	if (!functionType)
+	else
 	{
-		m_errorReporter.typeError(_functionCall.location(), "Type is not callable");
-		_functionCall.annotation().type = make_shared<TupleType>();
-		return false;
+		TypeChecker::typeCheckNonABIDecodeFunctionLike(_functionCall, _functionType);
```

Comment:
I think it would be better to give this a more generic name - "NonABIDecodingFunctionLike" does not sound like it would cover 99% of the cases :)

Reply:
I have renamed the method.